jdaviz/core/tests/test_data_menu.py

```diff
@@ -49,3 +50,34 @@ def test_data_menu_toggles(specviz_helper, spectrum1d):
     selected_data_items = app._viewer_item_by_id('specviz-0')['selected_data_items']
     assert selected_data_items[data_ids[0]] == 'visible'
     assert selected_data_items[data_ids[1]] == 'mixed'
+
+
+def test_visibility_toggle(imviz_helper, image_2d_wcs):
+    arr = np.ones((10, 10))
+    ndd = NDData(arr, wcs=image_2d_wcs)
+
+    imviz_helper.load_data(ndd, data_label='no_results_data')
+    app = imviz_helper.app
+    iv = app.get_viewer('imviz-0')
+
+    # regression test for https://github.com/spacetelescope/jdaviz/issues/1723
+    # test to make sure plot options (including percentile) stick when toggling visibility
+    selected_data_items = app._viewer_item_by_id('imviz-0')['selected_data_items']
```

Comment:
Huh, so this use case cannot be tested with a more public API? I can't say I understand what is really going on here. It would be better if we can test it as described in the original ticket, "Load one image, set the stretch percentile to something that is not 95%, load another image. The stretch gets reset to 95%."

Also you don't really need WCS to test this. You can load the same plain number array twice. What matter is the layer percentile setting.

Reply:
Maybe something like this:

```python
import numpy as np
from numpy.testing import assert_allclose

from jdaviz import Imviz

def test_visibility_toggle(imviz_helper):
    a = np.arange(4).reshape((2, 2))
    imviz_helper.load_data(a, data_label='image_1')
    imviz_helper.default_viewer.cuts = '90%'
    assert_allclose(imviz_helper.default_viewer.cuts, (0.15000000000000002, 2.8499999999999996))

    imviz_helper.load_data(a, data_label='image_2')
    imviz_helper.default_viewer.blink_once()  # Goes back to the first image (you need to double check if this is needed)
    assert_allclose(imviz_helper.default_viewer.cuts, (0.15000000000000002, 2.8499999999999996))
```

This code currently fails on `main` like this:

```
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 2 / 2 (100%)
Max absolute difference: 0.075
Max relative difference: 0.5
 x: array([0.075, 2.925])
 y: array([0.15, 2.85])
```

Reply:
The existing test triggers the code where the problem actually existed (literally checking the boxes, we confirmed offline that the report was for toggling the visibility not calling `load_data` from the API).  Blinking currently does trigger the same bug, but the implementation for blinking may change in the future, so for the sake of regression testing, I think it makes more sense to use non-public API here.  As far as using astrowidgets vs plot options to confirm they have not reverted, I don't see any advantage to one over the other 🤷 

I did remove the WCS as that definitely isn't needed - I had just copied that from another test since I'm not too familiar with all the imviz test fixtures.

Reply:
From a user perspective, I would be more interested in testing that the user workflow won't be broken in future changes. I don't care as much if the `app._viewer_item_by_id('imviz-0')['selected_data_items']` behavior changes or not, as long as the user workflow still works as expected. I guess it comes down to the philosophy of that this test is really testing.

Reply:
this does test the more direct user workflow of checking the boxes in the data menu UI (there just isn't a nice user-friendly API for that yet).  Blinking or changing layer visibility manually would still trigger this same logic, but relies on observers on the visibility state and that logic flow could change in the future.  I'd still like to keep this test since its the most direct way to trigger the code-block that had the bug, but if you think either of those other two workflows (blinking, manually setting layer visibility) warrants separate tests, we can do that.

Reply:
Yeah, por que no los dos? So...

* Add my suggestion as a second test.
* Add comments in your test so someone not you understands what it is supposed to be doing.

Sounds good?

Reply:
I'm trying to avoid too much redundancy.  The previous test (`test_data_menu_toggles`) already tests that a change to `layer.visible` updates the data menu.  This current test then tests that changes in the data menu will not reset the percentile as a regression test (I think that's an important distinction - this isn't something that would normally be tested if it wasn't for the fact that there used to be a bug in the logic that caused the `add_data_to_viewer` to be re-called - it's not anything fancy, just an overlooked if-statement).

I'm not sure the best way to write a blink test that decouples it from the data-menu.  I tried migrating your suggestion into `test_tools.test_blink`, but I think that just really clutters up that test and adds confusion.  The existing blink test effectively already ensures that at least the _top layer_ has been updated accordingly.  Currently we know this is done by setting the visibility state, which is a case already covered in the other existing tests.

I did modify the comment in the test to explain that it is designed to test changes to the checkbox in data menu being handled correctly (and that points back to the issue and here, so I'm not too worried about a lack of bookkeeping).

Reply:
While I'm just about to mark my approval since this test *does* test the bug, I just wanted to put I think I'm generally in favor of @pllim's thought of why not both? Fair, it's redundant, but I don't necessarily see redundancy in tests has a huge issue, especially in examples like this where it shows different entrypoints to the same behavior. For instance, we test app.load_data lots of times via all our configs independently testing their viz.load_data methods, but since they have different entrypoints, it checks not only the lower level methods, but also the different pathways to get there.

Granted this is a more proactive approach to testing, rather than need-based, so I'm happy to get this in and not have it hold up the review

Reply:
#1742 introduces a nicer (almost public, but not technically public since its at the app-level) API and updates the tests to make use of that.